src/serialization/serialization.rs

```diff
@@ -0,0 +1,412 @@
+        // therefore be about 5e7 (50 million) at most.
+        length.checked_mul(9)
+    }
+}
+
+/// Errors that can happen during deserialization.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum DeserializeError {
+    /// An i/o operation failed.
+    IoError,
+    /// The cookie (first 4 bytes) did not match that for any supported format.
+    InvalidCookie,
+    /// The histogram uses features that this implementation doesn't support (yet), so it cannot
+    /// be deserialized correctly.
+    UnsupportedFeature,
+    /// A count exceeded what can be represented in the chosen counter type.
+    UnsuitableCounterType,
+    /// The histogram instance could not be created because the serialized parameters were invalid
+    /// (e.g. lowest value, highest value, etc.)
+    InvalidParameters,
+    /// The current system's pointer width cannot represent the encoded histogram.
+    UsizeTypeTooSmall,
+    /// The encoded array is longer than it should be for the histogram's value range.
+    EncodedArrayTooLong
+}
+
+impl std::convert::From<std::io::Error> for DeserializeError {
+    fn from(_: std::io::Error) -> Self {
+        DeserializeError::IoError
+    }
+}
+
+/// Deserializer for all supported formats.
+/// Deserializers are intended to be re-used for many histograms.
+pub struct Deserializer {
+    payload_buf: Vec<u8>
+}
+
+impl Deserializer {
+    /// Create a new deserializer.
+    pub fn new() -> Deserializer {
+        Deserializer {
+            payload_buf: Vec::new()
+        }
+    }
+
+    /// Deserialize an encoded histogram.
+    pub fn deserialize<T: Counter, R: Read>(&mut self, reader: &mut R)
+                                            -> Result<Histogram<T>, DeserializeError> {
+        // TODO benchmark minimizing read calls by reading into a fixed-size header buffer
+
+        let cookie = reader.read_u32::<BigEndian>()?;
+
+        if cookie != V2_COOKIE {
+            return Err(DeserializeError::InvalidCookie);
+        }
+
+        let payload_len = reader.read_u32::<BigEndian>()?.to_usize()
+            .ok_or(DeserializeError::UsizeTypeTooSmall)?;
+        let normalizing_offset = reader.read_u32::<BigEndian>()?;
+        if normalizing_offset != 0 {
+            return Err(DeserializeError::UnsupportedFeature);
+        }
+        let num_digits = reader.read_u32::<BigEndian>()?.to_u8()
+            .ok_or(DeserializeError::InvalidParameters)?;
+        let low = reader.read_u64::<BigEndian>()?;
+        let high = reader.read_u64::<BigEndian>()?;
+        let int_double_ratio = reader.read_f64::<BigEndian>()?;
+        if int_double_ratio != 1.0 {
+            return Err(DeserializeError::UnsupportedFeature);
+        }
+
+        let mut h = Histogram::new_with_bounds(low, high, num_digits)
+            .map_err(|_| DeserializeError::InvalidParameters)?;
+
+        if payload_len > self.payload_buf.len() {
+            self.payload_buf.resize(payload_len, 0);
```

Comment:
We want it to only grow, never shrink. The vec here is a dumb buffer that only exists to be big enough to take an appropriately-sized slice of below. If the buffer is too big here, it might not be the next time...